pkg/handler/metatile.go

```diff
@@ -100,120 +132,216 @@ func MetatileHandler(
-		// thing into memory.
-		storageReadStart := time.Now()
-		bodySize, err := io.Copy(buf, storageResp.Body)
-		reqState.Duration.StorageRead = time.Since(storageReadStart)
+		responseData, err := extractVectorTileFromMetatile(reqState, bufferManager, parseResult, metatileResponseData)
 		if err != nil {
-			logger.Error(log.LogCategory_StorageError, "Failed to read storage body: %#v", err)
 			http.Error(rw, err.Error(), http.StatusInternalServerError)
-			reqState.FetchState = state.FetchState_ReadError
 			reqState.ResponseState = state.ResponseState_Error
 			return
 		}
-		reqState.FetchState = state.FetchState_Success
-
-		storageBytes := buf.Bytes()
-		reqState.FetchSize.BodySize = bodySize
-		reqState.FetchSize.BytesLength = int64(len(storageBytes))
-		reqState.FetchSize.BytesCap = int64(cap(storageBytes))
-
-		headers := rw.Header()
-		headers.Set("Content-Type", parseResult.ContentType)
-		if lastMod := storageResp.LastModified; lastMod != nil {
-			// important! we must format times in an HTTP-compliant way, which
-			// apparently doesn't match any existing Go time format string, so the
-			// recommended way is to switch to UTC and use the format string that
-			// the net/http package exposes.
-			lastModifiedFormatted := lastMod.UTC().Format(http.TimeFormat)
-			headers.Set("Last-Modified", lastModifiedFormatted)
-			reqState.StorageMetadata.HasLastModified = true
-		}
-		if etag := storageResp.ETag; etag != nil {
-			headers.Set("ETag", *etag)
-			reqState.StorageMetadata.HasEtag = true
-		}
 
-		metatileReaderFindStart := time.Now()
-		reader, formatSize, err := tile.NewMetatileReader(offset, bytes.NewReader(storageBytes), bodySize)
-		reqState.Duration.MetatileFind = time.Since(metatileReaderFindStart)
-		if err != nil {
-			logger.Error(log.LogCategory_MetatileError, "Failed to read metatile: %#v", err)
-			http.Error(rw, err.Error(), http.StatusInternalServerError)
-			reqState.IsZipError = true
-			reqState.ResponseState = state.ResponseState_Error
-			return
-		}
-		// make sure to close zip file reader
-		defer reader.Close()
-		reqState.ResponseSize = int(formatSize)
-		headers.Set("Content-Length", fmt.Sprintf("%d", formatSize))
-
-		rw.WriteHeader(http.StatusOK)
-		reqState.ResponseState = state.ResponseState_Success
-		respWriteStart := time.Now()
-		_, err = io.Copy(rw, reader)
-		reqState.Duration.RespWrite = time.Since(respWriteStart)
+		// Copy some of the metatile response data over to the vector tile response data so that it is properly cachedVecResp
+		responseData.ETag = metatileResponseData.ETag
+		responseData.LastModified = metatileResponseData.LastModified
+
+		err = writeVectorTileResponse(reqState, rw, responseData)
 		if err != nil {
+			// TODO Context cancellation might happen here?
 			logger.Error(log.LogCategory_ResponseError, "Failed to write response body: %#v", err)
-			reqState.IsResponseWriteError = true
+			// Still want to set the cache in this case
 		}
+
+		// Cache the response
```

Comment:
where do you SetMetaTile?

Reply:
Good catch. I hadn't committed that yet because I was debugging what I thought was a race condition with the metatile set. Turned out it was just not tracking hit/miss reporting correctly. Fixed in 9a95dfc.